atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/snapshot/DefaultKeyValueSnapshotReader.java

```diff
@@ -0,0 +1,383 @@
+                    // earlier timestamp.
+                    keysToReload.put(key, value.getTimestamp());
+                    metricRecorder.recordFilteredTransactionCommittingAfterOurStart(tableRef);
+                } else {
+                    // The value has a commit timestamp less than our start timestamp, and is visible and valid.
+                    if (value.getContents().length != 0) {
+                        resultsCollector.add(Maps.immutableEntry(key, transformer.apply(value)));
+                        keysAddedBuilder.add(key);
+                    }
+                }
+            }
+        }
+        Set<Cell> keysAddedToResults = keysAddedBuilder.build();
+
+        if (!keysToDelete.isEmpty()) {
+            // if we can't roll back the failed transactions, we should just try again
+            if (!rollbackFailedTransactions(tableRef, keysToDelete, commitTimestamps)) {
+                return Futures.immediateFuture(getRemainingResults(rawResults, keysAddedToResults));
+            }
+        }
+
+        if (!keysToReload.isEmpty()) {
+            // TODO (jkong): Consider removing when a decision on validating pre-commit requirements mid-read is made
+            return Futures.transform(
+                    transactionKeyValueService.getAsync(tableRef, keysToReload),
+                    nextRawResults -> {
+                        boolean allPossibleCellsReadAndPresent = nextRawResults.size() == keysToReload.size();
+                        readSnapshotValidator.throwIfPreCommitRequirementsNotMetOnRead(
+                                tableRef, startTimestampSupplier.getAsLong(), allPossibleCellsReadAndPresent);
+                        return getRemainingResults(nextRawResults, keysAddedToResults);
+                    },
+                    MoreExecutors.directExecutor());
+        }
+        return Futures.immediateFuture(ImmutableMap.of());
+    }
+
+    private Map<Cell, Value> getRemainingResults(Map<Cell, Value> rawResults, Set<Cell> keysAddedToResults) {
+        Map<Cell, Value> remainingResults = new HashMap<>(rawResults);
+        remainingResults.keySet().removeAll(keysAddedToResults);
+        return remainingResults;
+    }
+
+    private LongSet getStartTimestampsForValues(Iterable<Value> values) {
+        return LongSets.immutable.withAll(Streams.stream(values).mapToLong(Value::getTimestamp));
+    }
+
+    private ListenableFuture<LongLongMap> getCommitTimestamps(TableReference tableRef, LongIterable startTimestamps) {
+        return commitTimestampLoader.getCommitTimestamps(tableRef, startTimestamps);
+    }
+
+    /**
+     * This will attempt to rollback the passed transactions.  If all are rolled back correctly this method will also
+     * delete the values for the transactions that have been rolled back.
+     *
+     * @return false if we cannot roll back the failed transactions because someone beat us to it
+     */
+    private boolean rollbackFailedTransactions(
```

Comment:
What's your longer term plan for this class? Do you intend to keep rollback e.t.c in this class, or move it out to some other new abstraction?

Reply:
Also, do you plan on eventually moving tests from SnapshotTransaction that are relevant to this class into a test class just for KVSR?

Reply:
A q for duplication - do you want to leave a TODO here? Unless you've already prepared the work for it.

Reply:
The tests thing could be useful, though dependent on bandwidth 😬